icu4c/source/common/mlbe.cpp

```diff
@@ -144,96 +131,75 @@ int32_t MlBreakEngine::divideUpRange(UText *inText, int32_t rangeStart, int32_t
+        featureList[listLength].append(u"UW4:", 4).append(w4);
+        listLength++;
     }
     if (w5 != INVALID) {
-        arr[0] = w5;
-        concatChar(u"UW5:", arr, length, featureList[listLength++], status);
+        featureList[listLength].append(u"UW5:", 4).append(w5);
+        listLength++;
     }
     if (w6 != INVALID) {
-        arr[0] = w6;
-        concatChar(u"UW6:", arr, length, featureList[listLength++], status);
+        featureList[listLength].append(u"UW6:", 4).append(w6);
+        listLength++;
     }
-    length = 2;
     if (w2 != INVALID && w3 != INVALID) {
-        arr[0] = w2;
-        arr[1] = w3;
-        concatChar(u"BW1:", arr, length, featureList[listLength++], status);
+        featureList[listLength].append(u"BW1:", 4).append(w2).append(w3);
+        listLength++;
     }
     if (w3 != INVALID && w4 != INVALID) {
-        arr[0] = w3;
-        arr[1] = w4;
-        concatChar(u"BW2:", arr, length, featureList[listLength++], status);
+        featureList[listLength].append(u"BW2:", 4).append(w3).append(w4);
+        listLength++;
     }
     if (w4 != INVALID && w5 != INVALID) {
-        arr[0] = w4;
-        arr[1] = w5;
-        concatChar(u"BW3:", arr, length, featureList[listLength++], status);
+        featureList[listLength].append(u"BW3:", 4).append(w4).append(w5);
+        listLength++;
     }
-    length = 3;
     if (w1 != INVALID && w2 != INVALID && w3 != INVALID) {
-        arr[0] = w1;
-        arr[1] = w2;
-        arr[2] = w3;
-        concatChar(u"TW1:", arr, length, featureList[listLength++], status);
+        featureList[listLength].append(u"TW1:", 4).append(w1).append(w2).append(w3);
+        listLength++;
     }
     if (w2 != INVALID && w3 != INVALID && w4 != INVALID) {
-        arr[0] = w2;
-        arr[1] = w3;
-        arr[2] = w4;
-        concatChar(u"TW2:", arr, length, featureList[listLength++], status);
+        featureList[listLength].append(u"TW2:", 4).append(w2).append(w3).append(w4);
+        listLength++;
     }
     if (w3 != INVALID && w4 != INVALID && w5 != INVALID) {
-        arr[0] = w3;
-        arr[1] = w4;
-        arr[2] = w5;
-        concatChar(u"TW3:", arr, length, featureList[listLength++], status);
+        featureList[listLength].append(u"TW3:", 4).append(w3).append(w4).append(w5);
+        listLength++;
     }
     if (w4 != INVALID && w5 != INVALID && w6 != INVALID) {
-        arr[0] = w4;
-        arr[1] = w5;
-        arr[2] = w6;
-        concatChar(u"TW4:", arr, length, featureList[listLength++], status);
-    }
-    if (U_FAILURE(status)) {
-        return;
+        featureList[listLength].append(u"TW4:", 4).append(w4).append(w5).append(w6);
+        listLength++;
     }
+
     int32_t score = fNegativeSum;
     for (int32_t j = 0; j < listLength; j++) {
-        UnicodeString key(featureList[j]);
-        if (fModel.containsKey(key)) {
-            score += (2 * fModel.geti(key));
-        }
+        score += (2 * fModel.geti(featureList[j]));
```

Comment:
Please add a comment above this line.
```
// Because when the key doesn't exist geti(key) returns 0  and 2 * 0 = 0,
//  the following line will not be changed, so we can skip checking fModel
// includes key featureList[j] or not.
```

Reply:
Done